src/appleseed-max-impl/appleseedinteractive/interactiverenderercontroller.h

```diff
@@ -80,6 +85,106 @@ class CameraObjectUpdateAction
     renderer::Project&                                m_project;
 };
 
+class MaterialUpdateAction
+  : public ScheduledAction
+{
+  public:
+    MaterialUpdateAction(
+        renderer::Project&      project,
+        const IAppleseedMtlMap& material_map)
+      : m_material_map(material_map)
+      , m_project(project)
+    {
+    }
+
+    void update() override
+    {
+        renderer::Assembly* assembly = m_project.get_scene()->assemblies().get_by_name("assembly");
+        for (const auto& mtl : m_material_map)
+        {
+            renderer::Material* material = assembly->materials().get_by_name(mtl.second.c_str());
+
+            if (material)
+                assembly->materials().remove(material);
+
+            assembly->materials().insert(
+                mtl.first->create_material(
+                *assembly,
+                mtl.second.c_str(),
+                false,
+                GetCOREInterface()->GetTime()));
+        }
+    }
+
+  private:
+    IAppleseedMtlMap        m_material_map;
+    renderer::Project&      m_project;
+};
+
+class AssignMaterialAction
+  : public ScheduledAction
+{
+  public:
+    AssignMaterialAction(
+        renderer::Project&      project,
+        const MaterialMap&      material_map,
+        const InstanceMap&      instances)
+      : m_material_map(material_map)
+      , m_project(project)
+      , m_instances(instances)
+    {
+    }
+
+    void update() override
+    {
+        renderer::Assembly* assembly = m_project.get_scene()->assemblies().get_by_name("assembly");
+        for (const auto& mtl : m_material_map)
+        {
+            renderer::Material* material = assembly->materials().get_by_name(mtl.second.c_str());
+
+            if (material == nullptr)
+            {
+                const auto mtl_name =
+                    make_unique_name(assembly->materials(), wide_to_utf8(mtl.first->GetName()) + "_mat");
+
+                IAppleseedMtl* appleseed_mtl =
+                    static_cast<IAppleseedMtl*>(mtl.first->GetInterface(IAppleseedMtl::interface_id()));
+                if (appleseed_mtl == nullptr)
+                    continue;
+
+                const size_t index = assembly->materials().insert(
+                    appleseed_mtl->create_material(
+                    *assembly,
+                    mtl_name.c_str(),
+                    false,
+                    GetCOREInterface()->GetTime()));
+                material = assembly->materials().get_by_index(index);
+            }
+
+            for (const auto& obj_instance : m_instances)
+            {
+                auto* instance = obj_instance.first;
+                auto& obj_info = obj_instance.second;
+
+                auto front_mapping = instance->get_front_material_mappings();
```

Comment:
Replace `auto` by real type.

Keep a const ref instead of making a copy (if possible; I don't know).